forgerock-authenticator/src/main/java/org/forgerock/android/auth/PushResponder.java

```diff
@@ -255,6 +229,74 @@ public void onFailure(@NotNull Call call, @NotNull IOException e) {
         }
     }
 
+    /**
+     * Used to respond an authentication request from a given message
+     *
+     * @param pushNotification The push notification object.
+     * @param approved The approval response
+     * @param payload The payload
+     * @param listener Listener for receiving the HTTP call response code.
+     */
+    private void performAuthentication(@NonNull PushNotification pushNotification,
+                                       boolean approved,
+                                       Map<String, Object> payload,
+                                       final @NonNull FRAListener<Void> listener)
+            throws MalformedURLException, JSONException, JOSEException {
+
+        // Check if notification has been approved
+        if(!pushNotification.isPending()) {
+            listener.onException(new PushMechanismException("PushNotification is not in a" +
+                    " valid status to authenticate; either PushNotification has already been" +
+                    " authenticated or expired."));
+            return;
+        }
+
+        // Get authentication endpoint
+        String endpoint = pushNotification.getPushMechanism().getAuthenticationEndpoint();
+        URL url = new URL(endpoint);
+
+        // Build request
+        OkHttpClient okHttpClient = getOkHttpClient(url);
+        Request request = buildRequest(url,
+                pushNotification.getAmlbCookie(),
+                pushNotification.getPushMechanism().getSecret(),
+                pushNotification.getMessageId(),
+                payload);
+
+        // Invoke URL
+        okHttpClient.newCall(request).enqueue(new okhttp3.Callback() {
+            @Override
+            public void onResponse(@NotNull Call call, @NotNull Response response) {
+                Logger.debug(TAG, "Response from server: \n" + response.toString());
+                // Check if operation succeed
+                if(response.code() == 200) {
+                    // Update notification status
+                    pushNotification.setPending(false);
+                    if(approved)
+                        pushNotification.setApproved(true);
+                    else
+                        pushNotification.setApproved(false);
+                    // Persist status
+                    if(!storageClient.setNotification(pushNotification))
+                        listener.onException(new PushMechanismException("Push Authentication " +
```

Comment:
For code style, I would prefer to put { } for if statement, even with only one line